full-stack-tests/presentation/src/learning-snippets/Hierarchies.test.ts

```diff
@@ -24,305 +28,1613 @@ describe("Learning Snippets", () => {
+              ruleType: RuleTypes.ChildNodes,
+              condition: `ParentNode.IsOfClass("Model", "BisCore")`,
+              specifications: [{
+                specType: ChildNodeSpecificationTypes.RelatedInstanceNodes,
+                relationshipPaths: [{
+                  relationship: { schemaName: "BisCore", className: "ModelContainsElements" },
+                  direction: RelationshipDirection.Forward,
+                  targetClass: { schemaName: "BisCore", className: "GeometricElement3d" },
+                }],
+              }],
+            }],
+          };
+          // __PUBLISH_EXTRACT_END__
+          printRuleset(ruleset);
+
+          // Verify that correct Model Elements are returned, grouped by class
+          const modelNodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+          expect(modelNodes).to.have.lengthOf(1).and.to.containSubset([{
+            key: { instanceKeys: [{ className: "BisCore:PhysicalModel" }] },
+          }]);
+
+          const elementClassGroupingNodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset, parentKey: modelNodes[0].key });
+          expect(elementClassGroupingNodes).to.have.lengthOf(2).and.to.containSubset([{
+            label: { displayValue: "Physical Object" },
+          }, {
+            label: { displayValue: "TestClass" },
+          }]);
+
+          const elementNodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset, parentKey: elementClassGroupingNodes[0].key });
+          expect(elementNodes).to.have.lengthOf(2).and.to.containSubset([{
+            key: { instanceKeys: [{ className: "Generic:PhysicalObject" }] },
+          }, {
+            key: { instanceKeys: [{ className: "Generic:PhysicalObject" }] },
+          }]);
+        });
+
+      });
+
+      describe("CustomNodeSpecification", () => {
+
+        it("uses `type` attribute", async () => {
+          // __PUBLISH_EXTRACT_START__ Hierarchies.CustomNodeSpecification.Type.Ruleset
+          // The ruleset has a root node specification that returns a single custom node with specified parameters.
+          const ruleset: Ruleset = {
+            id: "example",
+            rules: [{
+              ruleType: RuleTypes.RootNodes,
+              specifications: [{
+                specType: ChildNodeSpecificationTypes.CustomNode,
+                type: "T_MY_NODE",
+                label: "My Node",
+              }],
+            }],
+          };
+          // __PUBLISH_EXTRACT_END__
+          printRuleset(ruleset);
+
+          // __PUBLISH_EXTRACT_START__ Hierarchies.CustomNodeSpecification.Type.Result
+          // Verify that node with correct type is returned
+          const nodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+          expect(nodes).to.have.lengthOf(1).and.to.containSubset([{
+            key: { type: "T_MY_NODE" },
+          }]);
+          // __PUBLISH_EXTRACT_END__
```

Comment:
In this specific case (the `type` attribute), maybe it would be better to add a child node rule to the ruleset with condition `ParentNode.Type="T_MY_NODE"`? That's a more common use case for this `type` attribute.

However, there are similar situations elsewhere (e.g. the `description` attribute test), where I create a snippet with `expect` function. I see what you mean, but I don't see how switching to `assert` changes anything - it also comes from a library. I also don't think exposing a snippet with `expect` is a problem - the snippet is readable enough even without understanding what the individual function does. The reader can easily treat it as pseudo-code, if not familiar with chai.

Reply:
I would prefer an example showing how custom node is referenced in a ruleset and displaying the resulting tree image.

Code snippets that parrot our tests should have no place in documentation. We can display code, but it needs to be presented in a way that people will actually use. Making an assertion is a plausible use case that is also straightforward to adapt to other contexts.

Assertion functions have been around for decades in almost all programming languages, so everyone understands how and why they are used. Chai.expectations.are.way.harder.to.read.compared.to.simple.C-like.access.pattern.

Reply:
> I would prefer an example showing how custom node is referenced in a ruleset and displaying the resulting tree image.

Will make the change.

> Code snippets that parrot our tests should have no place in documentation. We can display code, but it needs to be presented in a way that people will actually use.

The thing is that presentation library isn't directly tied to UI - you can make similar requests purely on the backend. So displaying a UI screenshot is not always possible, especially when different UI components may handle the same attribute in different ways. So in those cases I'm adding a code snippet showing that some attribute is carried over to the result and from there it's up to the consumer to choose what he wants to do.

> Assertion functions have been around for decades and are present in almost all programming languages, so everyone understands how and why they are used. Chai.expectations.are.way.harder.to.read.compared.to.simple.C-like.access.pattern.

I surely hope that our consumers are smart enough to understand what `expect(x).to.not.be.undefined` or similar means. As I said, even if you don't know chai, this piece can be treated as pseudo-code. I don't want our tests, let alone a single test, to mix different types of assertions.

Reply:
Your arguments are very weak, especially because Hirarchies.test.ts already uses asserts in multiple places.

You completely missed my points, but whatever.

Reply:
I'm resolving this conversation because there is more effort being spent derailing the thread than addressing the reason behind not following a well established documentation convention.